lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -2038,6 +2038,63 @@ MutableOperandRange cudaq::cc::ConditionOp::getMutableSuccessorOperands(
   return getResultsMutable();
 }
 
+//===----------------------------------------------------------------------===//
+// IfOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// const prop `cc.if($true)` and `cc.if($false)`
+struct EraseConstIf : public OpRewritePattern<cudaq::cc::IfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  static bool hasQuantumAllocations(Block &block) {
+    for (auto &op : block) {
+      for (auto &region : op.getRegions())
+        for (auto &b : region)
+          if (hasQuantumAllocations(b))
+            return true;
+
+      if (auto alloc = dyn_cast<quake::AllocaOp>(&op))
+        return true;
+    }
+    return false;
+  }
+
+  LogicalResult matchAndRewrite(cudaq::cc::IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    auto cond = ifOp.getCondition();
+    if (auto constCond = cond.getDefiningOp<arith::ConstantOp>()) {
+      if (auto attr = dyn_cast<BoolAttr>(constCond.getValue())) {
+        auto value = attr.getValue();
+        auto block =
+            value ? ifOp.getThenEntryBlock() : ifOp.getElseEntryBlock();
+
+        // Prevent quantum allocations from moving to an outer scope.
+        if (block && hasQuantumAllocations(*block))
+          return failure();
```

Comment:
Was there an example where this happened? If so, the IR has bugs.

*Any* allocations must be either in a basic block owned by the func::FuncOp, cc::LambdaCreateOp, or a cc::ScopeOp. So this check isn't needed for valid IR.

Reply:
Replaced this pattern with RewriteIf from lower-to-cfg